tools/closed_by_commits.py

```diff
@@ -0,0 +1,351 @@
+
+        results: dict[Issue, set[PullRequest]] = defaultdict(set)
+        for pr_node in pr_nodes:
+            pr = PullRequest(
+                number=pr_node["number"],
+                title=pr_node["title"].strip(),
+            )
+            pr_comments = pr_node.get("comments", {}).get("nodes", [])
+            for node in [pr_node, *pr_comments]:
+                pr.czo_urls.update(self._extract_czo_urls(node.get("body")))
+            pr.czo_urls.update(search_czo_for_number(pr.number))
+
+            # Get issues from PR metadata
+            issues_dict = {}
+            issue_nodes = pr_node.get("closingIssuesReferences", {}).get("nodes", [])
+            for issue_node in issue_nodes:
+                if issue_node is None:
+                    continue
+
+                issue = Issue(
+                    number=issue_node["number"],
+                    title=issue_node["title"].strip(),
+                )
+                issue_comments = issue_node.get("comments", {}).get("nodes", [])
+                for node in [issue_node, *issue_comments]:
+                    issue.czo_urls.update(self._extract_czo_urls(node.get("body")))
+                issue.czo_urls.update(search_czo_for_number(issue.number))
+                issues_dict[issue.number] = issue
+
+            # Extract additional issue numbers from commit message
+            for issue_num in self._extract_issue_numbers(commit_message):
+                if issue_num not in issues_dict:
+                    maybe_issue = self._fetch_issue(issue_num)
+                    if maybe_issue is None:
+                        continue
+                    issues_dict[issue_num] = maybe_issue
+
+            for issue in issues_dict.values():
+                results[issue].add(pr)
+
+        if not pr_nodes:
+            for issue_num in self._extract_issue_numbers(commit_message):
+                maybe_issue = self._fetch_issue(issue_num)
+                if maybe_issue is None:
+                    continue
+                results[maybe_issue].update()
+
+        return results
+
+    def analyze_range(self, base: str, head: str) -> list[Issue]:
+        repository = self.github.get_repo("zulip/zulip")
+        comparison = repository.compare(base, head)
+        commit_shas = [commit.sha for commit in comparison.commits]
+        logger.info("Found %d commits", len(commit_shas))
+
+        issue_to_prs: dict[Issue, set[PullRequest]] = defaultdict(set)
+        for i, sha in enumerate(commit_shas, 1):
+            logger.debug("Processing commit %d/%d: %s", i, len(commit_shas), sha)
+
+            for issue, prs in self.get_issues_for_commit(sha).items():
+                issue_to_prs[issue].update(prs)
+
+        final_issues = []
+        for issue, prs in issue_to_prs.items():
+            for pr in prs:
+                issue.czo_urls.update(pr.czo_urls)
+
+            issue.closed_by_prs = sorted(prs, key=lambda p: p.number)
+            final_issues.append(issue)
```

Comment:
nit: seems like `final_issues` is the same as `list(issue_to_prs.keys())`; could replace it below with `issue_to_prs` or `issue_to_prs.keys()`